pyccl/halos/pk_4pt.py

```diff
@@ -505,3 +585,976 @@ def _logged_output(*arrs, log):
+        T^{1h}_{13}({\\bf k_1},-{\\bf k_1},{\\bf k_2},-{\\bf k_2}),
+
+
+    Args:
+        cosmo (:class:`~pyccl.core.Cosmology`): a Cosmology object.
+        hmc (:class:`HMCalculator`): a halo model calculator.
+        k (float or array_like): comoving wavenumber in Mpc^-1.
+        a (float or array_like): scale factor.
+        prof (:class:`~pyccl.halos.profiles.HaloProfile`): halo
+            profile (corresponding to :math:`u_1` above.
+        prof2 (:class:`~pyccl.halos.profiles.HaloProfile`): halo
+            profile (corresponding to :math:`u_2` above. If `None`,
+            `prof` will be used as `prof2`.
+        prof3 (:class:`~pyccl.halos.profiles.HaloProfile`): halo
+            profile (corresponding to :math:`v_1` above. If `None`,
+            `prof` will be used as `prof3`.
+        prof4 (:class:`~pyccl.halos.profiles.HaloProfile`): halo
+            profile (corresponding to :math:`v_2` above. If `None`,
+            `prof3` will be used as `prof4`.
+        prof12_2pt (:class:`~pyccl.halos.profiles_2pt.Profile2pt`):
+            a profile covariance object returning the 2-point
+            moment of `prof`, `prof2`. If `None`, the default second moment
+            will be used, corresponding to the products of the means of each
+            profile.
+        prof34_2pt (:class:`~pyccl.halos.profiles_2pt.Profile2pt`):
+            same as `prof34_2pt` for `prof3` and `prof4`.
+        p_of_k_a (:class:`~pyccl.pk2d.Pk2D`): a `Pk2D` object to
+            be used as the linear matter power spectrum. If `None`, the power
+            spectrum stored within `cosmo` will be used.
+
+    Returns:
+        float or array_like: integral values evaluated at each
+        combination of `k` and `a`. The shape of the output will
+        be `(N_a, N_k, N_k)` where `N_k` and `N_a` are the sizes of
+        `k` and `a` respectively. The ordering is such that
+        `output[ia, ik2, ik1] = T(k[ik1], k[ik2], a[ia])`
+        If `k` or `a` are scalars, the corresponding dimension will
+        be squeezed out on output.
+    """
+    a_use = np.atleast_1d(a)
+    k_use = np.atleast_1d(k)
+
+    # Check inputs
+    prof, prof2, prof3, prof4, prof12_2pt, prof34_2pt, _, _ = \
+        _allocate_profiles2(prof, prof2, prof3, prof4, prof12_2pt, prof34_2pt,
+                            None, None)
+
+    # Power spectrum
+    def get_pk(p_of_k_a):
+        if isinstance(p_of_k_a, Pk2D):
+            def pkf(sf):
+                return p_of_k_a(k_use, sf, cosmo)
+        elif (p_of_k_a is None) or (str(p_of_k_a) == 'linear'):
+            def pkf(sf):
+                return cosmo.get_linear_power()(k_use, sf)
+        elif str(p_of_k_a) == 'nonlinear':
+            def pkf(sf):
+                return cosmo.get_nonlin_power()(k_use, sf)
+        else:
+            raise TypeError("p_of_k_a must be `None`, \'linear\', "
+                            "\'nonlinear\' or a `Pk2D` object")
+        return pkf
```

Comment:
We could call this just once outside the loop, right? (just setting the function, I mean)

Reply:
Done